raster-foundry/raster-foundry
# Datasource band support

## Overview

This PR uses the `bands` of datasources to display more accessible band names in the UI.
It:
  - adds a new component `band-select` which encapsulates the logic of properly displaying a datasource's bands and allowing selection.
  - updates all of the color-composite UI to use the new component
  - updates lab input nodes to use the new component along with style adjustments for alignment
  - adds cached datasource removal when datasource is updated

### Checklist

~- [ ] Styleguide updated, if necessary~
~- [ ] Swagger specification updated, if necessary~
~- [ ] Symlinks from new migrations present or corrected for any new migrations~

### Demo

![image](https://user-images.githubusercontent.com/2442245/35080718-694a20be-fbdd-11e7-964b-05e3c4524ee9.png)

![image](https://user-images.githubusercontent.com/2442245/35080737-77ad6cba-fbdd-11e7-9056-a95897213996.png)

![image](https://user-images.githubusercontent.com/2442245/35080106-8f666cac-fbd9-11e7-9331-6021bdbefb30.png)

![image](https://user-images.githubusercontent.com/2442245/35080083-70da6752-fbd9-11e7-9726-e05dbd205871.png)

![image](https://user-images.githubusercontent.com/2442245/35080046-31138824-fbd9-11e7-953b-edc41f26a599.png)


### Notes + Questions

@designmatty @notthatbreezy 

**Editor: Assuming a 'Natural Color' composite makes sense by default** - _unaddressed_
When we create a datasource, we also add a default color composite named 'Natural Color' (I've changed that to 'Default' for now) which has a band ordering of (R: 0, G: 1, B: 2). I don't necessarily think that this makes sense. Not only may the datasource have less than 3 bands but the band ordering may very well be different.

It may make more sense to do a single band map (R: 0, G: 0, B: 0) by default since we are always guaranteed one band.

**Lab: allowing a user to set an input node's band without having a project defined** - _addressed_
Previously, a user could set the band of an input node without having defined a project to use as the source. Now, a band cannot be selected unless a project is first selected.

If a different project is selected, the merged bands of the previous and current projects are compared. The band selection is only preserved when they match exactly; otherwise, the band selection is removed.


## Testing Instructions

 * Edit a project that has only scenes from a datasource with defined bands and test out the single band and custom color composites

 * Edit a project that has only scenes from a datasource _without_ defined bands and make sure you can't adjust single or custom color composites. Make sure you are prompted to edit that datasource.

 * Test band selection in the lab.
    - When you switch projects and the projects have mismatched bands, the band selection should be cleared.
    - When you switch projects and the projects have the same bands (i.e. are from the same datasource, the band selection is preserved.

Closes #2512
